Memory planning (#3)
Menooker/llvm-project: mlir/docs/MemrefMergeAlloc.md

considering much about the overall memory reuse across Operators within a
sub-graph (or partition).

The current implementation of Bufferization and memref pass pipeline focuses on
copy-avoidance and in-place reusing of the memory. Consider a computation graph
of 4 layers of matmul sharing the same weight:
```mlir
func.func @mlp(%x: tensor<128x128xf32>, %y: tensor<128x128xf32>) -> tensor<128x128xf32> {
   %a0 = tensor.empty() : tensor<128x128xf32>
   %a = linalg.matmul ins(%x, %y: tensor<128x128xf32>, tensor<128x128xf32>) outs(%a0: tensor<128x128xf32>) -> tensor<128x128xf32>
   %b0 = tensor.empty() : tensor<128x128xf32>
   %b = linalg.matmul ins(%a, %y: tensor<128x128xf32>, tensor<128x128xf32>) outs(%b0: tensor<128x128xf32>) -> tensor<128x128xf32>
   %c0 = tensor.empty() : tensor<128x128xf32>
   %c = linalg.matmul ins(%b, %y: tensor<128x128xf32>, tensor<128x128xf32>) outs(%c0: tensor<128x128xf32>) -> tensor<128x128xf32>
   %d0 = tensor.empty() : tensor<128x128xf32>
   %d = linalg.matmul ins(%c, %y: tensor<128x128xf32>, tensor<128x128xf32>) outs(%d0: tensor<128x128xf32>) -> tensor<128x128xf32>
   return %d : tensor<128x128xf32>

> }

Reviewer: Maybe we can update this example with different buffer size for the tensors (like `a0` and `c0`, etc) or even complex scenario like control flow for better demonstrating the benefit of this proposal?

Author: I would like a simple example to show the core idea of the pass: memory reuse, merged alloc, replace by `memref.view`

   return %d0 : memref<128x128xf32>
}
```

Without further optimizations, 3 temp buffers will be allocated at the runtime
for these tensors. However, as we can see in the IR, the buffer `a0` is no
longer used when buffer `c0` is allocated. So buffer `c0` can reuse the memory
buffer of buffer `a0`, to reduce the memory size footprint and improve the
locality.

An observation of the current bufferization and memref passes is that they do
not consider the memory buffer planning - to reuse the buffer/memref for less
total size and better locality.

## Merge-alloc pass
An optimization pass has been introduced to consolidate multiple allocations
(`memref.alloc` ops) into a single `memref.alloc` op and each static-shaped
`memref.alloc` op will be transformed into a "slice" from the `single allocated
buffer` with `memref.view` and some compile-time decided `offsets`. This
optimization works on `memref` instead of `tensor` ops, so it should be executed
after bufferization pass, and before adding buffer deallocation ops.

While merging the memory allocations, the transform should consider the lifetime

> of each allocated `memref`s. By lifetime, we mean the range of time when an
> memref allocated from `memref.alloc` is actively used. The references on `view`s

Reviewer: ```suggestion
of each allocated `memref`. By lifetime, we mean the range of time when a
memory allocated from `memref.alloc` is actively used. The references on `view`
```

Author: by `memref`s, I mean plural form of memref, and I think I'd better not put `s` in `memref`.

Reviewer: Seems you're mixing the usage of `allocation`, `memory`, and `buffer` in the doc, I think most of them are identity right?
The comment here is mainly for line 73, memref->memroy.

Author: OK, changed.

of a "base" `memref` should contribute to the lifetime of the "base". A later
`memref.alloc` should consider to reuse the memory of a previously allocated
memref, if the lifetime of these two does not overlap. The transform will
perform the "reusing" of memory by setting the `offset` of the later
`memref.view` to a position within the memory range of a previous allocation's

> `memref.view` on the `single allocated buffer`.

Reviewer: Do you mean `memref.alloc`? or trying to say the relationship between the current `memref.view` and the previous one?

Author: It is a typo. Thanks for finding

mergable memref.alloc with views on the merged allocations with the offsets
calculated in the `MemorySchedule`. This step is abstracted in a
`MemoryMergeMutatorFunc` function.

```c++
using MemoryMergeMutatorFunc = std::function<LogicalResult(
    Operation *toplevel, Operation *scope, const MemorySchedule &,
    const MergeAllocationOptions &)>;
```

## Tick-based Implementation for merge-alloc

A tick-based implementation of merge-alloc in provided by default. The basic
idea of the tick-based allocation merging is that

1. Each of the operations in a function is assigned a "tick". An operation with
   a smaller tick is expected to be executed before one with a larger tick
2. Collect the first referenced tick and the last referenced tick for each
   mergeable allocation. If a buffer is referenced in loops and branches,
   special handling is needed.
3. For each allocation scope, linearize the first referenced tick and the last
   referenced tick of mergeable allocations inside of it into a single linear
   timeline
4. Use a "static-memory-planner" to handle the linear timeline

Limitations of Tick-based merge-alloc:

>  * only contiguous, static shaped and identical layout memrefs are considered.

Reviewer: Seems these are the common limitation for this fwk, not only this default implementation?

Author: In theory, we don't need limit that. If an implementation can magically schedule the dynamic shaped tensor (e.g. by considering the additionaly restrictions of the shapes), it is possible to schedule the buffers.

   Others are disregarded
 * only `RegionBranchOpInterface` or `LoopLikeOpInterface` operations are
   allowed to access memref inside the operations' children regions. Other
   operaions containing regions should not access memref inside. Otherwise, a
   pass error could occur.

### Basic concepts

In the context of tick-based merge-alloc, mergeable alloction and allocation
scope are defined as follows

#### Mergeable alloction

The pass should only consider to merge a `memref.alloc` only if
 * the ownership of the memref does not escape from the function. That is, the
   current function is responsible to alloc and dealloc this memref
 * and the allocated memref is contiguous and has static shape and identical
   layout.

>  * and memref is in the default memory space (this restriction may be removed in

Reviewer: What do you mean by "default memory space"?

Author: `memory space` is a term in memref and LLVM, usually to mark if a pointer is for host or for device.

Reviewer: I see, thanks.

   the future)

In tick-based merge-alloc, we call these `memref.alloc` **mergeable**
allocations.

The memrefs passed by function arguments, or returned by the function will be
untouched by this optimization.

#### Allocation scopes

The transformation first needs to identify the allocation scopes, which are mlir
operaions containing non-zero regions, and
 * implementing `AutomaticAllocationScope`
 * and is not `scf.for` (allocations in an `scf.for` can be hoisted to parent
   `AutomaticAllocationScope`)

For example, below is an example IR of a function with nested `scf.forall` ops.

```mlir
func.func @mlp(...) { // <---- alloc scope 1
   scf.for(...) { // <---- NOT an alloc scope!

>       // allocation inside will be merge to alloc scope 1 above

Reviewer: What's the rewrite order of nested alloc scope? How do you handle the new buffer which is hoisted out from `scf.for`?

Author: The hoisting is "one-shot" - it is decided by the `getAllocationScope`. There will be no second run of hoisting.

appropriate free chunk, and split the chunk into two chunks - one for the memory
range for the allocation, and another the remaining free memory range. On an
"free" event, the memory planner marks the memory chunk as "free". If the
neighbouring memory chunks are also "free", the planner will further merge the
neighbouring free chunks into a larger free chunk.

A improvement of static memory planner over runtime memory allocators is that,
if a "free" memory chunk has smaller size than an allocation size, the memory
planner is allowed to "extend" the size of the "free" memory chunk to match the
allocation. It is not possible for runtime memory allocators, because extending
a memory chunk involves moving the memory addresses of the previously allocated
memory. However, in our compile-time memory planner, all the allocations are
logical, and the offsets of the allocated memory ranges can always be adjusted
by a later allocation. This improvement helps to reduce the issue of memory
fragmentation.

On an "alloc" event, the memory planner needs to choose one candidate from all
"free" memory chunks. A memory chunk that is recently free'd is considered "hot"

> in cache. In the default configuration (when `no-consider-locality` option is

Reviewer: The option name should be updated as well.

Author: Thanks. Updated.

not specified to the merge-alloc pass), static memory planner considers both
cache-locality and the degree of matching of allocation size and the chunk size

> for each free memory chunks, with a simple cost-model. With

Reviewer: "cost-model" might be too heavy? use "heuristic rule" instead?

Author: do we need to rename the pass option as well?

Reviewer: How about `locality-and-size` and `size-only`?
